src/core/reactor.cc

```diff
@@ -3995,7 +3995,9 @@ void smp::configure(const smp_options& smp_opts, const reactor_options& reactor_
     if (thread_affinity) {
         smp::pin(allocations[0].cpu_id);
     }
-    memory::configure(allocations[0].mem, mbind, hugepages_path);
+    if (smp_opts.memory_allocator == memory_allocator::seastar) {
+        memory::configure(allocations[0].mem, mbind, hugepages_path);
```

Comment:
original_malloc_func is never written. We can make it const (in a follow up).

Furthermore, is_reactor_thread will not be set since we don't call memory::configure().

Reply:
If `original_malloc_func` never changes, why is this if then:
```c++
void* allocate(size_t size) {
    if (!is_reactor_thread) {
        if (original_malloc_func) {                                             <<<<<<< This one
            alloc_stats::increment(alloc_stats::types::foreign_mallocs);
            return original_malloc_func(size);
        }
        // original_malloc_func might be null for allocations before main
        // in constructors before original_malloc_func ctor is called
        init_cpu_mem();
    }
```

Reply:
because allocate() can be called before the initializer of original_malloc_func() is called.

("never changes" is wrong - the code never tries to change it, but because allocate() is so low-level it can see it changing when the initializer initializes it)

Reply:
There's even a comment!

Reply:
That's _exactly_ my question -- this set avoids calling `memory::configure()` together with `init_cpu_mem()`, but still `allocate()` can "cancel" this, call `init_cpu_mem()` and fall-back to non-original malloc. Is that OK?

Reply:
You're right. is_reactor_thread is used here to check for two conditions: 1. this is an alien thread. 2. we're running before initialization.

It's not pretty, but it's necessary for things to work. The initialization of original_malloc_func needs allocations, so we're calling init_cpu_mem to provide some way for things to work in that small window before the initializer gets to work. It will only allocate 32MB.

So my previous explanation was wrong, but the patch is still correct.

It may have been better not to override __libc_malloc() and initialize original_malloc_func = __libc_malloc, avoiding the window. But that exposes us to internal libc calls to __libc_malloc bypassing our allocator.